tools/registry.py

```diff
@@ -0,0 +1,291 @@
+      else:
+        raise RegistryException(
+            f"Version {module.version} for module {module.name} already exists.")
+
+    p = self.root.joinpath("modules", module.name, module.version)
+    p.mkdir()
+
+    # Create MODULE.bazel
+    module_dot_bazel = p.joinpath("MODULE.bazel")
+    if module.module_dot_bazel:
+      shutil.copy(module.module_dot_bazel, module_dot_bazel)
+    else:
+      deps = "\n".join(
+          f"bazel_dep(name = \"{name}\", version = \"{version}\")"
+          for name, version in module.deps)
+      with module_dot_bazel.open("w") as f:
+        f.write(self._MODULE_BAZEL.format(
+            module.name, module.version,
+            module.compatibility_level) + deps)
+        f.write("\n")
+
+    # Create source.json & copy patch files to the registry
+    source = {
+        "url": module.url,
+        "integrity": integrity(download(module.url)),
+    }
+    if module.strip_prefix:
+      source["strip_prefix"] = module.strip_prefix
+
+    patch_dir = p.joinpath("patches")
+    if module.patches or module.build_file:
+      patch_dir.mkdir()
+      source["patches"] = {}
+      source["patch_strip"] = module.patch_strip
+
+    if module.patches:
+      for s in module.patches:
+        patch = pathlib.Path(s)
+        source["patches"][patch.name] = integrity(read(patch))
+        shutil.copy(patch, patch_dir)
+
+    # Turn additional BUILD file into a patch
+    if module.build_file:
+      build_file_content = pathlib.Path(module.build_file).open().readlines()
+      build_file = "a/" * module.patch_strip + "BUILD.bazel"
+      patch_content = difflib.unified_diff(
+          [], build_file_content, "/dev/null", build_file)
+      patch_name = "add_build_file.patch"
+      patch = patch_dir.joinpath(patch_name)
+      with patch.open("w") as f:
+        f.writelines(patch_content)
+      source["patches"][patch_name] = integrity(read(patch))
+
+    json_dump(p.joinpath("source.json"), source)
+
+    # Create presubmit.yml file
+    presubmit_yml = p.joinpath("presubmit.yml")
+    if module.presubmit_yml:
+      shutil.copy(module.presubmit_yml, presubmit_yml)
+    else:
+      platforms = {
+          "centos7": {},
+          "debian10": {},
+          "ubuntu2004": {},
+          "macos": {},
+          "windows": {},
+      }
+      for _, config in platforms.items():
+        if module.build_targets:
+          config["build_targets"] = module.build_targets.copy()
+        if module.test_targets:
+          config["test_targets"] = module.test_targets.copy()
+      with presubmit_yml.open("w") as f:
+        yaml.dump({"platforms": platforms}, f)
+
+    # Add new version to metadata.json
+    metadata_path = self.root.joinpath("modules", module.name,
+                                       "metadata.json")
+    metadata = json.load(metadata_path.open())
+    metadata["versions"].append(module.version)
+    metadata["versions"] = list(set(metadata["versions"])).sort()
```

Comment:
this sort() call might be oversimplistic. we might one day need to implement the version sorting algorithm in Python too. (also for the presubmit check.) Maybe just leave a TODO for now since that sounds like a lot of work.

Reply:
Good point! Added a TODO here.
Also realized sort() actually does the sorting in place and doesn't return anything. Fixed the bug.